### ISSUE TITLE
How to get distance between two images after training on customize dataset？

### ISSUE THREAD

**Issue by whalefa1I** [leondgarse/Keras_insightface#44]
您好！
我按照readme中的说明，使用train.py获得了checkpoints下的若干文件：例如kereas_mobilenet_emore.h5等。
接下来应当如何获得某一图片的embedding呢？谢谢

**Comment by leondgarse** [comment]
参照 [Real time face recognition with Android Problem #7](https://github.com/leondgarse/Keras_insightface/issues/7)
```py
iaa = np.zeros([1, 112, 112, 3]) # 图片 1
ibb = np.ones([1, 112, 112, 3]) # 图片 2

mm = keras.models.load_model('checkpoints/kereas_mobilenet_emore.h5') # 加载 basic_model 模型
eaa = mm(iaa) # 图片 1 的特征
ebb = mm(ibb) # 图片 2 的特征
print(eaa.shape, ebb.shape)
# (1, 256) (1, 256)

from sklearn.preprocessing import normalize
eea = normalize(eaa) # 归一化
eeb = normalize(ebb)
print((eea ** 2).sum(), (eeb ** 2).sum())
# 1.0 1.0

print(np.dot(eea, eeb.T)) # 计算 cos 距离相似度
# [[0.38431713]]
```

**Comment by whalefa1I** [comment]
开始我是这么做的，然后报错出现unknown layer：NormDense，是需要通过keras的custom_objec把这个config加进去吗

**Comment by whalefa1I** [comment]
哦去掉NormDense，懂了，我回去再啃论语，谢谢夫子

**Comment by leondgarse** [comment]
皮

**Comment by whalefa1I** [comment]
对了哥，如果训练数据格式为这样的话：
.               # dataset folder
├── Anglelababy           # person folder
│   ├── 100.jpg # face image
│   ├── 101.jpg # face image
│   └── 102.jpg # face image
├── Yamy           # person folder
│   ├── 111.jpg
│   ├── 112.jpg
│   └── 113.jpg
├── xjb
│   ├── 707.jpg
│   ├── 708.jpg
│   └── 709.jpg
可以修改训练代码满足这样的目录结构吗

**Comment by whalefa1I** [comment]
是不是改个image_classes_rule就行了

**Comment by leondgarse** [comment]
是的，用这个 [class ImageClassesRule_map](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L13)
可以改 [data.py#L47](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L47)
```py
- image_classes_rule = default_image_classes_rule
+ image_classes_rule = ImageClassesRule_map()
```

**Comment by whalefa1I** [comment]
![error](https://user-images.githubusercontent.com/32239722/130566776-ab70e788-2a82-45c1-913f-bc8173847481.PNG)
这个error您见过吗

**Comment by whalefa1I** [comment]
是不是版本问题哦

**Comment by whalefa1I** [comment]
o 这个项目是tf1.X的是不是！

**Comment by leondgarse** [comment]
是版本问题，目前我用的是 `TF2.6.0`

**Comment by leondgarse** [comment]
`TF 1.13` 中的参数是 `from_generator(generator, output_types, output_shapes=None, args=None)`，可以改改试试
```py
# 286
- ds = tf.data.Dataset.from_generator(self.image_shuffle_gen, output_signature=output_signature)
+ ds = tf.data.Dataset.from_generator(self.image_shuffle_gen, (tf.string, tf.int64), (tf.TensorShape([]), tf.TensorShape([]))
```
这个我也没使用过，而且如果是 `TF 1` 的话可能还有其他问题。。

**Comment by whalefa1I** [comment]
我用basic_model.h5去eval_folder了自己的数据集，看起来效果还行嗷？
![image](https://user-images.githubusercontent.com/32239722/130576005-0c7bc1da-5e93-4f64-a321-2e89b0050852.png)
但是我手动去查看这个数据集中的两个图片的时候
![image](https://user-images.githubusercontent.com/32239722/130576657-5c490914-7398-4b02-a913-010632c410d7.png)
shape好像不太对哎，而且结果好像也不太对，是哪里有问题吗
![image](https://user-images.githubusercontent.com/32239722/130576742-f1b2f89b-06ea-4314-aaab-0483616d3e39.png)
我还用了basic_model.h5去试了试，shape好像也不对？
![image](https://user-images.githubusercontent.com/32239722/130576932-96c40529-c6c5-4a2d-9b6a-63e3755f91ef.png)
训练就用的
![image](https://user-images.githubusercontent.com/32239722/130577072-0f9e07c6-c661-4a23-b245-a08f4f5aec1c.png)

**Comment by leondgarse** [comment]
1. 结果好坏得看数据集难度，最好有相同数据集上的对比，比如 readme 中的 `mobilenet` 再 `IJBB` 数据集上  `TAR@FAR=1e-4` 位置是 `0.91889`.
2. 你这个 `embedding_model` 是删多了，就是移除的层多了，现在输出的应该是 `flatten` 层的结果。训练过程中保存的 `basic_model` 是可以直接用的，不用再移除层，`NormDense` 层是训练模型中的。
3. `basic_model.h5` 输出的 shape 是正确的。
4. 另外的话，计算相似度的部分没有做归一化，就是这部分
  ```py
  from sklearn.preprocessing import normalize
  eea = normalize(eaa) # 归一化
  eeb = normalize(ebb)
  ```

**Comment by leondgarse** [comment]
啊，因为你指定的 `eval_paths` 是空的，所以训练过程中没有保存 `basic_model` 模型。。
你可以在 [train.py#L408](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/train.py#L408) 这里加一个 `callback`，类似
```py
self.callbacks.append(keras.callbacks.LambdaCallback(on_epoch_end=lambda epoch, logs=None: self.basic_model.save("aa_epoch{}.h5".format(epoch))))
```
指定每个 epoch 结束保存一下，或者自己在训练结束后 `tt.basic_model.save('aa.h5')` 保存一下

**Comment by whalefa1I** [comment]
> 是的，用这个 [class ImageClassesRule_map](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L13)
> 可以改 [data.py#L47](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L47)
> 
> ```python
> - image_classes_rule = default_image_classes_rule
> + image_classes_rule = ImageClassesRule_map()
> ```

这是不是这个意思：

```
class ImageClassesRule_map:
    def __init__(self, dir, dir_rule="*", excludes=[]):
        - raw_classes = [os.path.basename(ii) for ii in glob2.glob(os.path.join(dir, dir_rule))]
        + raw_classes = [os.path.basename(ii) for ii in glob2.glob(os.path.join(os.path.dirname(dir), dir_rule))]
        # 毕竟 image_classes = [image_classes_rule(ii) for ii in image_names]
        # 这里循环的是每张图片的路径

        self.raw_classes = sorted([ii for ii in raw_classes if ii not in excludes])
        self.classes_2_indices = {ii: id for id, ii in enumerate(self.raw_classes)}
        self.indices_2_classes = {vv: kk for kk, vv in self.classes_2_indices.items()}

    def __call__(self, image_name):
        raw_image_class = os.path.basename(os.path.dirname(image_name))
        return self.classes_2_indices[raw_image_class]
```

**Comment by leondgarse** [comment]
不是，`data.py#L47` 的意思是 `data.py` 的第 47 行，点链接过去就是那里

**Comment by whalefa1I** [comment]
> 是的，用这个 [class ImageClassesRule_map](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L13)
> 可以改 [data.py#L47](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L47)
> 
> ```python
> - image_classes_rule = default_image_classes_rule
> + image_classes_rule = ImageClassesRule_map()
> ```
改成我的image_classes变成了ImageClassesRule_map的list，debug的时候根本没有到__call__那边
然后classes_ = np.max(image_classes) -1

> 不是，`data.py#L47` 的意思是 `data.py` 的第 47 行，点链接过去就是那里

- image_classes_rule = default_image_classes_rule
+ image_classes_rule = ImageClassesRule_map(data_path)
那这里就是这样

**Comment by whalefa1I** [comment]
> 不是，`data.py#L47` 的意思是 `data.py` 的第 47 行，点链接过去就是那里

那这里是不是应该是这样？


> 啊，因为你指定的 `eval_paths` 是空的，所以训练过程中没有保存 `basic_model` 模型。。
> 你可以在 [train.py#L408](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/train.py#L408) 这里加一个 `callback`，类似
> 
> ```python
> self.callbacks.append(keras.callbacks.LambdaCallback(on_epoch_end=lambda epoch, logs=None: self.basic_model.save("aa_epoch{}.h5".format(epoch))))
> ```
> 
> 指定每个 epoch 结束保存一下，或者自己在训练结束后 `tt.basic_model.save('aa.h5')` 保存一下



> 不是，`data.py#L47` 的意思是 `data.py` 的第 47 行，点链接过去就是那里
那这里是这样？

```
- image_classes_rule = default_image_classes_rule
+ image_classes_rule = ImageClassesRule_map(data_path)
```

**Comment by leondgarse** [comment]
没看懂你的改动，这两处改动的话
- [data.py#L47](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L47)，意思是在 `data.py` 的 47 行
  ```py
  - image_classes_rule = default_image_classes_rule  # 删掉这行，不要前面减号
  + image_classes_rule = ImageClassesRule_map()  # 增加这行，不要前面加号
  ```
- [train.py#L408](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/train.py#L408)，意思是在 `train.py` 的 408 行添加下面这行
  ```py
  self.callbacks.append(keras.callbacks.LambdaCallback(on_epoch_end=lambda epoch, logs=None: self.basic_model.save("aa_epoch{}.h5".format(epoch))))
  ```

**Comment by whalefa1I** [comment]
> 没看懂你的改动，这两处改动的话
> 
> * [data.py#L47](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L47)，意思是在 `data.py` 的 47 行
>   ```python
>   - image_classes_rule = default_image_classes_rule  # 删掉这行，不要前面减号
>   + image_classes_rule = ImageClassesRule_map()  # 增加这行，不要前面加号
>   ```
> * [train.py#L408](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/train.py#L408)，意思是在 `train.py` 的 408 行添加下面这行
>   ```python
>   self.callbacks.append(keras.callbacks.LambdaCallback(on_epoch_end=lambda epoch, logs=None: self.basic_model.save("aa_epoch{}.h5".format(epoch))))
>   ```

![image](https://user-images.githubusercontent.com/32239722/130608161-f5469950-85a2-4d57-bd7e-57ec4dea23bf.png)
- image_classes_rule = default_image_classes_rule
+ image_classes_rule = ImageClassesRule_map(data_path)
嗯嗯我看懂了哥，我是说这边需要加一个data_path吧？(好想微信给你转个红包

**Comment by leondgarse** [comment]
哦，没注意到这个 data_path，是的，是有个 data_path。。
TripletLoss 需要 positive + negative 数据，同一个类中的 positive 太少的话可能不是太好，所以加了这个限制，你可以在 [data.py#L259](https://github.com/leondgarse/Keras_insightface/blob/ad81c00a1e2f35db9ccbcdb674b7f455400583c5/data.py#L259) 把 `4` 改小点，去掉这个限制。`image_per_class` 是 `train.Train` 的参数，控制实际使用的单个 class 内数量。

**Comment by whalefa1I** [comment]
训练准确率啥的90+但是距离还是分不出来，正例和负例的距离区别不是很明显，只能多找点数据训练吗，还是说处理计算距离可以有其他办法去判断两个人是不是一个人呢

**Comment by leondgarse** [comment]
- 如果数据预处理正确的话，正常来讲一般是可以区分的。过程大概 `人脸检测 -> 对齐 -> [-1, 1] -> 模型输出 embedding -> 归一化 -> 计算距离`，[video_test.py](https://github.com/leondgarse/Keras_insightface/blob/master/video_test.py) 等测试脚本中都是类似的过程。
- 也可以参照这里 [Colab efficientnetV2_basic_test.ipynb](https://colab.research.google.com/drive/1vmAEfF9tUgK2gkrS5qVftadTyUcX343D?usp=sharing) 最后面 `Keras insightface test` 部分，可以看到不同人之间相似度最高也是 `< 0.3` 的。一般 cosine 距离可以选取 `0.6` 作为阈值，`> 0.6` 判定为同一个人。距离计算上一般就是余弦距离或者欧氏距离。
- 如果以上都没有问题，有可能是训练数据的问题了，训练的模型泛化不好。最有效的方式还是增强数据集，其他可以考虑数据增强 / dropout / stochastic depth 等防止过拟合方式。
